src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,9 +197,149 @@ mender_storage_init(void) {
+                     depl_logs_buffer.f_sector_cnt,
+                     (depl_logs_buffer.f_sector_cnt - 1) * mender_storage_nvs_handle.sector_size);
+#endif /* CONFIG_MENDER_DEPLOYMENT_LOGS */
+
     return MENDER_OK;
 }
 
+#ifdef CONFIG_MENDER_DEPLOYMENT_LOGS
+mender_err_t
+mender_storage_deployment_log_append(const char *msg, size_t msg_size) {
+    int          result;
+    mender_err_t ret = MENDER_OK;
+
+    msg_size = MIN(DEPL_LOGS_MAX_MSG_LEN + 1, msg_size);
+
+    const struct flash_area *fap;
+    if (0 != (result = flash_area_open(FIXED_PARTITION_ID(MENDER_STORAGE_LABEL), &fap))) {
+        mender_log_error("Unable to open the Mender storage flash area");
+        return MENDER_FAIL;
+    }
+
+    struct fcb_entry entry = { 0 };
+    result                 = fcb_append(&depl_logs_buffer, msg_size, &entry);
+    if (-ENOSPC == result) {
+        if (0 != (result = fcb_rotate(&depl_logs_buffer))) {
+            mender_log_error("Failed to rotate the deployment logs FCB");
+            /* if (0 != (result = fcb_append_to_scratch(&depl_logs_buffer))) { */
+            /*     mender_log_error("Failed to start using the scratch sector for the deployment logs FCB"); */
+            ret = MENDER_FAIL;
+            goto END;
+        }
+        const char rotation_msg[] = "<wrn> ------- DEPLOYMENT LOGS ROTATED -------";
+        result                    = fcb_append(&depl_logs_buffer, sizeof(rotation_msg), &entry);
+        if (0 != (result = flash_area_write(fap, entry.fe_sector->fs_off + entry.fe_data_off, rotation_msg, sizeof(rotation_msg)))) {
+            mender_log_error("Failed to write message to the FCB entry [%d]", -result);
+            ret = MENDER_FAIL;
+            goto END;
+        }
+        result = fcb_append_finish(&depl_logs_buffer, &entry);
+        if (0 != result) {
+            mender_log_error("Failed to finish append of a rotation entry to the deployment logs FCB");
+            ret = MENDER_FAIL;
+            goto END;
+        }
+
+        result = fcb_append(&depl_logs_buffer, msg_size, &entry);
+    }
+    if (0 != result) {
+        mender_log_error("Failed to append a new entry to the deployment logs FCB");
+        ret = MENDER_FAIL;
+        goto END;
+    }
+    /* else success, proceed */
+    if (0 != (result = flash_area_write(fap, entry.fe_sector->fs_off + entry.fe_data_off, msg, msg_size))) {
+        mender_log_error("Failed to write message to the FCB entry [%d]", -result);
+        ret = MENDER_FAIL;
+        goto END;
+    }
+    result = fcb_append_finish(&depl_logs_buffer, &entry);
+    if (0 != result) {
+        mender_log_error("Failed to finish append of a new entry to the deployment logs FCB");
+        ret = MENDER_FAIL;
+        goto END;
+    }
+
+END:
+    flash_area_close(fap);
+
+    return ret;
+}
+
+mender_err_t
+mender_storage_deployment_log_walk(MenderDeploymentLogVisitor visitor_fn, void *ctx) {
```

Comment:
It is the first time I see the visitor pattern in C :heart_eyes:

Reply:
Then you should see "my" `JsonWalk()`: https://github.com/NorthernTechHQ/libntech/blob/master/libutils/json.h#L543-L568 :sweat_smile: